File: baekjoon/2293/solution.js
function solution(n, k, coins) {
  let prev = [...new Array(k + 1)].map((_, i) => (i === 0 ? 1 : 0));
  let current;
  for (let i = 1; i <= n; i++) {
    current = [...new Array(k + 1)].map((_, i) => (i === 0 ? 1 : 0));
    for (let j = 1; j <= k; j++) {
      const coin = coins[i - 1];
      current[j] = prev[j] + (j - coin >= 0 ? current[j - coin] : 0);
    }
    prev = current;
  }

  console.log(current[k]);
}

//////
////
// input

const fs = require("fs");
const isLocal = process.platform !== "linux";
const filePath = isLocal ? "./input.txt" : "/dev/stdin";
const input = fs.readFileSync(filePath).toString().trim();

const cases = input.split("\n").filter((item) => !!item);

let idx = 0;
while (idx < cases.length) {
  const [n, k] = cases[idx].split(" ").map((it) => +it);
  const offset = 1;
  const rows = cases.slice(idx + offset, idx + n + offset).map((item) => {
    return +item;
  });
  solution(n, k, rows);

  idx += n + offset;
}

File: baekjoon/2293/solution.py
import sys
sys.stdin = open("input.txt", "r") 

line = str(sys.stdin.readline())

[n, k] = map(lambda item: int(item), line.split(' '))

coins = []
for _ in range(n):
  coins.append(int(sys.stdin.readline()))

prev = list(map(
  lambda index_item: 1 if index_item[0] == 0 else 0,
  enumerate([0] * (k + 1))
))
current = []

for i in range(1, n + 1):
  current = list(map(
    lambda index_item: 1 if index_item[0] == 0 else 0,
    enumerate([0] * (k + 1))
  ))
  for j in range(1, k + 1):
    coin = coins[i - 1]
    current[j] = prev[j] + (current[j - coin] if j - coin >= 0 else 0)
  prev = current

print(current[k])
